### PR TITLE
Refactor worker for parallel processing and optimize data serialization

### PR DESCRIPTION
**Pull request by iuwqyir**
### TL;DR

Refactored worker package to improve performance and data handling for blockchain indexing.

### What changed?

- Introduced batch processing for blocks, logs, and traces
- Implemented parallel fetching of data using goroutines
- Refactored serialization functions to handle bulk data
- Updated data structures to better align with blockchain data models
- Removed direct storage dependencies from the worker
- Added support for configurable batch sizes for different data types

### How to test?

1. Run the indexer with different batch sizes for blocks, logs, and traces
2. Compare performance metrics before and after the changes
3. Verify that all data is correctly fetched and serialized
4. Test with different blockchain networks to ensure compatibility

### Why make this change?

This refactoring aims to significantly improve the indexer's performance by:

1. Reducing the number of RPC calls through batch processing
2. Utilizing parallel processing to speed up data fetching
3. Optimizing data structures for more efficient handling of blockchain data
4. Improving flexibility by allowing configurable batch sizes
5. Enhancing maintainability by separating concerns and reducing dependencies

These changes should result in faster indexing speeds, reduced resource consumption, and improved scalability of the indexer.

### CODE REVIEW

**Comment by iuwqyir** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/46?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#47** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/47?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#46** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/46?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#45** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/45?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#42** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/42?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#41** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/indexer/41?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @iuwqyir and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: internal/worker/worker.go [hunk 1]
```diff
@@ -27,98 +24,253 @@ type BlockResult struct {
...
+			Result: new(map[string]interface{}),
+		}
+	}
+
+	err := w.rpc.RPCClient.BatchCallContext(context.Background(), batch)
 	if err != nil {
-		return BlockResult{BlockNumber: blockNumber, Error: fmt.Errorf("error fetching logs for block %d: %v", blockNumber, err)}
+		for i := range blockResults {
+			blockResults[i].Error = err
+		}
```
Comment (AmineAfia) [inline internal/worker/worker.go:126]: This means if one batch element fails we fail all elements?

File: internal/worker/worker.go [hunk 2]
```diff
@@ -27,98 +24,253 @@ type BlockResult struct {
...
-		traces, err = w.fetchTraces(blockNumber)
-		if err != nil {
-			return BlockResult{BlockNumber: blockNumber, Error: fmt.Errorf("error fetching traces for block %d: %v", blockNumber, err)}
+	for i, elem := range batch {
+		blockResults[i].BlockNumber = blockNumbers[i]
+
+		if elem.Error != nil {
+			blockResults[i].Error = elem.Error
+		} else {
+			result := *elem.Result.(*map[string]interface{})
```
Comment (AmineAfia) [inline internal/worker/worker.go:136]: what does the `.(*map[string]interface{})` mean?

File: internal/worker/worker.go [hunk 3]
```diff
@@ -27,98 +24,253 @@ type BlockResult struct {
...
+			Method: "eth_getLogs",
+			Args:   []interface{}{map[string]string{"fromBlock": hexBlockNum, "toBlock": hexBlockNum}},
+			Result: new([]map[string]interface{}),
+		}
+	}
+
+	err := w.rpc.RPCClient.BatchCallContext(context.Background(), batch)
 	if err != nil {
-		log.Error().Err(err).Msg("Error querying ClickHouse")
+		for i := range logsResults {
```
Comment (AmineAfia) [inline internal/worker/worker.go:195]: same here, does this mean one failed block mean all batch should fail?

File: internal/worker/worker.go [hunk 4]
```diff
@@ -27,98 +24,253 @@ type BlockResult struct {
...
-	var result []map[string]interface{}
-	err := w.rpc.RPCClient.Call(&result, "trace_block", fmt.Sprintf("0x%x", blockNumber))
-	return result, err
+	return results
 }
 
-func (w *Worker) queryRows() {
-	// TODO: Implement this
-	/*rows, err := w.storage.DBStorage.Query(context.Background(), "SELECT version()")
+func (w *Worker) fetchLogsBatch(blockNumbers []*big.Int) []LogsFetchResult {
```
Comment (AmineAfia) [inline internal/worker/worker.go:177]: encapsulating this login in an extra function is agood idea, maybe it makes sense to generalize the function for batch fetching and use it for blocks and traces as well

File: internal/worker/worker.go [hunk 5]
```diff
@@ -27,98 +24,253 @@ type BlockResult struct {
...
+
+	results := make([]TracesFetchResult, 0, len(blockNumbers))
+	for batchResults := range resultsCh {
+		results = append(results, batchResults...)
+	}
+
+	return results
+}
+
+func (w *Worker) fetchTracesBatch(blockNumbers []*big.Int) []TracesFetchResult {
```
Comment (AmineAfia) [inline internal/worker/worker.go:244]: this code looks duplicated across blocs, transactions and traces. is it possible to generalize it?

**Review by AmineAfia — CHANGES_REQUESTED** [review]
lgtm, I just have some nit questions

File: internal/worker/worker.go (hunk 1 above)
Comment (iuwqyir) [reply internal/worker/worker.go:126]: No, this means if the batch request fails all together, we set the error for every block

File: internal/worker/worker.go (hunk 2 above)
Comment (iuwqyir) [reply internal/worker/worker.go:136]: it casts it to this type (basically an object with key value pairs)
Problem was that if I tried to use the types from go-ethereum, then batch requests did not work.
So I went for serializing them myself

File: internal/worker/worker.go (hunk 3 above)
Comment (iuwqyir) [reply internal/worker/worker.go:195]: Same response as to the other one, so resolving this

File: internal/worker/worker.go (hunk 4 above)
Comment (iuwqyir) [reply internal/worker/worker.go]: Yeah good idea, generalized the logic and it's much cleaner now

File: internal/worker/worker.go (hunk 5 above)
Comment (iuwqyir) [reply internal/worker/worker.go]: Done

File: internal/worker/worker.go (hunk 1 above)
Comment (AmineAfia) [reply internal/worker/worker.go]: The doc in the code is telling it differently. The function already sets the error object for the blocks that fails. the err object means one or more blocks failed. https://github.com/ethereum/go-ethereum/blob/80b529ea713d71a27282ae76c2741c8bc5486502/rpc/client.go#L395

Comment (iuwqyir) [reply internal/worker/worker.go]: That is how works. We send a batch request
* request works, all responses are successful -> we set the result for every block
* request works, but not all responses are successful -> we set results for successful blocks and Error for unsuccessful ones (the error is on the element of  `blockResults`)
* (this current case) request does not work and returns a single error -> we propagate that error for every block number

**Review by AmineAfia — APPROVED** [review]

**Comment by iuwqyir** [comment]
### Merge activity

* **Sep 25, 6:48 PM EDT**: @iuwqyir started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thirdweb-dev/indexer/46).

**Pull request merged**